src/connectionInfo/connectionManager.ts

```diff
@@ -0,0 +1,159 @@
+import { isIPv4 } from 'net';
+import * as vscode from 'vscode';
+import * as nls from 'vscode-nls';
+import { DEFAULT_HOST } from '../utils/constants';
+import { Disposable } from '../utils/dispose';
+import { SETTINGS_SECTION_ID, SettingUtil } from '../utils/settingsUtil';
+import { Connection, ConnectionInfo } from './connection';
+
+/**
+ * @description keeps track of all of the Connection objects and the info needed to create them (ie: initial ports).
+ */
+const localize = nls.loadMessageBundle();
+export class ConnectionManager extends Disposable {
+	private _initHttpPort;
+	private _initWSPort;
+	private _initHost: string;
+	private _connections: Map<vscode.Uri | undefined, Connection>; // undefined key means no workspace root
+
+	constructor(private readonly _extensionUri: vscode.Uri) {
+		super();
+
+		this._initHttpPort = SettingUtil.GetConfig(this._extensionUri).portNumber;
+		this._initWSPort = this._initHttpPort + 1;
```

Comment:
What if this port is in use?

Reply:
There is a check when the server is actually opened, which calls this. https://github.com/microsoft/vscode-livepreview/blob/d98f924b920543ef4de8df70c4581ab3d992333a/src/server/serverManager.ts#L408 It essentially tries to listen to a port and returns it if there's nothing on it. I do notice that it only gets called once and the ws server just takes the port after it, which may be wrong. I recall something about how the websocket server is able to be shared as long as the URL for it is different, but it seems that I don't check that. 

I'll create an issue for that specifically, but the HTTP port should never be taken up. If the initial port is taken up, then it tries the next one. Once the server initializes, it refreshes the preview if the preview uses that connection:
https://github.com/microsoft/vscode-livepreview/blob/d98f924b920543ef4de8df70c4581ab3d992333a/src/editorPreview/webviewComm.ts#L36

Reply:
Thanks, didn't see where that was happening